### ISSUE TITLE
And again about the non-English letters

### ISSUE THREAD

**Issue by shrv** [SpiderLabs/owasp-modsecurity-crs#220]
Suffered from a branch modseсurity. Question closer to this thread.

Good afternoon.
I apologize in advance for my bad English.
We use ModSecurity as WAF to our web resources. The version currently 2.7.0.

Charset of our pages utf-8. Since Russian resources, then the characters used in the URL or in the body of the query, refers to both the English and to the Russian alphabet.

What would be clear, with a real example of a query.
Consider for example the rules with ID 981173
SecRule ARGS_NAMES|ARGS|XML:/\* "([~!\@#\$\%^&_()-+\={}[]|:\;\"\'\´\’\‘`<>]._?){4,}" "phase:2,t:none,t:urlDecodeUni,block,id:'981173',rev:'2',ver:'OWASP_CRS/2.2.9',maturity:'9',accuracy:'8',msg:'Restricted SQL Character Anomaly Detection Alert - Total # of special characters exceeded',capture,logdata:'Matched Data: %{TX.1} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',tag:'OWASP_CRS/WEB_ATTACK/SQL_INJECTION',setvar:tx.anomaly_score=+%{tx.warning_anomaly_score},setvar:tx.sql_injection_score=+1,setvar:'tx.msg=%{rule.msg}',setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/RESTRICTED_SQLI_CHARS-%{matched_var_name}=%{tx.0}"

During experimentation, I broke it into three separate rules

SecRule ARGS_NAMES|ARGS|XML:/\* "([~!\@#\$\%^&_()-+\={}[]|:\;\"\'`<>]._?){4,}" "phase:2,t:none,t:urlDecodeUni,block,id:'901201',rev:'2',ver:'OWASP_CRS/2.2.9',maturity:'9',accuracy:'8',msg:'Restricted SQL Character Anomaly Detection Alert - Total # of special characters exceeded',capture,logdata:'Matched Data: %{TX.1} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',tag:'OWASP_CRS/WEB_ATTACK/SQL_INJECTION',setvar:tx.anomaly_score=+%{tx.warning_anomaly_score},setvar:tx.sql_injection_score=+1,setvar:'tx.msg=%{rule.msg}',setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/RESTRICTED_SQLI_CHARS-%{matched_var_name}=%{tx.0}"

SecRule ARGS_NAMES|ARGS|XML:/\* "(\´.*?){2,}" "phase:2,t:none,t:urlDecodeUni,block,id:'901202',rev:'2',ver:'OWASP_CRS/2.2.9',maturity:'9',accuracy:'8',msg:'Restricted SQL Character Anomaly Detection Alert - Total # of special characters exceeded',capture,logdata:'Matched Data: %{TX.1} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',tag:'OWASP_CRS/WEB_ATTACK/SQL_INJECTION',setvar:tx.anomaly_score=+%{tx.warning_anomaly_score},setvar:tx.sql_injection_score=+1,setvar:'tx.msg=%{rule.msg}',setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/RESTRICTED_SQLI_CHARS-%{matched_var_name}=%{tx.0}"

SecRule ARGS_NAMES|ARGS|XML:/\* "([\’\‘].*?){2,}" "phase:2,t:none,t:urlDecodeUni,block,id:'901203',rev:'2',ver:'OWASP_CRS/2.2.9',maturity:'9',accuracy:'8',msg:'Restricted SQL Character Anomaly Detection Alert - Total # of special characters exceeded',capture,logdata:'Matched Data: %{TX.1} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',tag:'OWASP_CRS/WEB_ATTACK/SQL_INJECTION',setvar:tx.anomaly_score=+%{tx.warning_anomaly_score},setvar:tx.sql_injection_score=+1,setvar:'tx.msg=%{rule.msg}',setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/RESTRICTED_SQLI_CHARS-%{matched_var_name}=%{tx.0}"

Divided into three rules based on the simple principle. The first rule in decoding all symbols are of the form \ххх, for example ~ is \x7E. In the second rule \xxx\xxx , for example ´ is \xC2\xB4. In a third \xxx\xxx\xxx, for example ’ is \xE2\x80\x99.

Now take the query, we are interested only body 
2_1=0&2_2=0&5_1_r=single&2_3=предприятиях по добыче и переработке.
предприятиях по добыче и переработке - this text contains only Russian letters. Any letter of the Russian alphabet has the form \xxx\xxx. Example р is \xD1\x80.

And see the following logs (worked the third rule)
ID Rule 901203
Rule Message: Restricted SQL Character Anomaly Detection Alert - Total # of special characters exceeded
Event: Pattern match "([\xe2\x80\x99\xe2\x80\x98].*?){2,}" at ARGS:2_3 
Data: Matched Data: \x80 found within ARGS:2_3: \xd0\xbf\xd1\x80\xd0\xb5\xd0\xb4\xd0\xbf\xd1\x80\xd0\xb8\xd1\x8f\xd1\x82\xd0\xb8\xd1\x8f\xd1\x85 \xd0\xbf\xd0\xbe \xd0\xb4\xd0\xbe\xd0\xb1\xd1\x8b\xd1\x87\xd0\xb5 \xd0\xb8 \xd0\xbf\xd0\xb5\xd1\x80\xd0\xb5\xd1\x8

Received a false alarm. We see that the rule, instead of what would as a pattern for comparison use \xE2\x80\x99, uses \xE2, \x80 as individual characters. It seems to me, is not correct.

Maybe it makes sense to consider the following scenario, I think it will reduce the number of false positives. Similar rules are divided into several. As described above. Introduces a new transformation functions t:length1, length2, etc. describes the length of the pattern with which to compare the values of the arguments, arguments? We also need to find ', which is exactly \ xE2 \ x80 \ x99, but not \ x80.

Further, the text of which seek coincidence divided into tighter length as the pattern, and each piece is compared with it so long as the length of the pattern pieces of equal length. For those from the original text rejected the first \ xxx element and then divided into pieces of equal length of the pattern. Repeat the comparison. Again reject \ xxx element. And so long as the length of the remaining piece of the original text will not be less than the length of the pattern.

Or maybe easier and the problem is not that it is not right to compare but not in the correct settings do I have?

**Comment by csanders-git** [comment]
You might not like this answer but here it goes...
Starting in version 3.0.0 (https://github.com/SpiderLabs/owasp-modsecurity-crs/tree/v3.0.0-dev) we removed that special chars rule. It was found to contain far to many false positives of the type you are describing. While a lot of the SQL logic stays the same, 3.0.0 relies heavily on the @detectSQLi which requires modsecurity 2.8 or greater.

If you find that you must use the current master version of OWASP CRS (older version's of ModSec that can't be updated), I would be more then happy to discuss other techniques for defending against the false positives you describe.

**Comment by shrv** [comment]
Yes, the month of April will be more accurate to use the current version 2.7.7. And maybe longer. In the near future I translate protection on two resources in the lock mode. And be updated to the latest version will be only after checking it on a test server.

Thank you for the correction of information in the rules. when I read the summary of the changes in the latest version, see this

**Comment by csanders-git** [comment]
Feel free to continue to reach out and stay involved in the project, we are always looking for issues like these. In general as far as other languages I myself never really get to see that side of ModSec. My recommendation might be to simply turn that rule off in your environment in the mean time until you can upgrade.... I'll give it some thought about how to better fix the issue over the course of this week.

**Comment by shrv** [comment]
Thank you for trying to help :).
Rules will not turn off, I used Anomaly Scoring Detection Mode and pair of false positives will not lead to the blocking request :)

**Comment by shrv** [comment]
Looked rules on the link that you sent.
I was very alarmed by setting
t:none,t:utf8toUnicode,t:urlDecodeUni

Let me explain
encodings are supported Russian letters are considered win-1251 and koi8-r

If I put in the file the following settings
SecUnicodeMapFile /ххх/ххх/ххх/ххх/modsec/unicode.mapping 20866
And apply a rule
t:none,t:utf8toUnicode,t:urlDecodeUni
then the query string containing the Russian letter is starting to look like this
\x1e?@. \x12! N 25-\x10\x1f\x1314-7 \x1e\x22 04.02.2015 
in the original
%D0%9E%D0%BF%D1%80.+%D0%92%D0%A1+N+25-%D0%90%D0%9F%D0%9314-7+%D0%9E%D0%A2+04.02.2015
in russian so
Опр. ВС N 25-АПГ14-7 ОТ 04.02.2015

using the win-1251 in unikode.mapping
line looks like %uxxxx%uxxxx and e.t.c.
Unfortunately there is now a real example, logs with them already deleted

**Comment by shrv** [comment]
@csanders-git 
good afternoon.
I've been thinking about the problem of checking non-english letters.

And it may make sense to look aside not Unicode, and punycode? What do you think? Or have a chance to make some "holes" in security?

**Comment by csanders-git** [comment]
what do you mean look aside?

**Comment by shrv** [comment]
I mean that is used to convert utf8 to Unicode
He proposed to consider the option utf8 in punycode

I'm here still thought about his idea. Remember, it is a stupid idea, nothing happens. :)

**Comment by facundofarias** [comment]
Is that 3.0.0 version already released it?

**Comment by csanders-git** [comment]
No we're waiting on a decent solution to this problem. Any help/commits would be greatly appreciated.

**Comment by vgvassilev** [comment]
@shrv Did the update to 3.0.0 solve any of your issues?

**Comment by shrv** [comment]
@vgvassilev No, not decided

**Comment by vgvassilev** [comment]
@shrv Thanks. I'd be happy to follow up and learn more if you decide to update.

**Comment by shrv** [comment]
@vgvassilev 
You do not understand me.
I upgraded to the specified version.
The update does not solve the problem.

**Comment by vgvassilev** [comment]
@shrv Thanks for the information. That is very unfortunate... :(

**Comment by csanders-git** [comment]
@shrv wanna hit me up offline we can see what changes need to be made to v3 to allow support for this we were considering a branch with Oniguruma, we're not sure how that will interact with the existing rules.

**Comment by shrv** [comment]
Good afternoon.
Now I have the core mod_sec v2.9.
A set of rules is a mix of the rules of version 2.2.9 and 3.0.

All the rules, which had a lot of false positives due to the Russian alphabet was replaced by the rule 3.0.
Unfortunately it did not help.

It briefly about the fact that I now have.

What would you like to know what my information you need and will be useful?

**Comment by dune73** [comment]
We know that the Core Rule Set does not handle non-latin character sets very well. CRS3 improves the situation, but the problem is nowhere really solved. For CRS3 we have cleaned out a lot of cruft and streamlined many rules, etc.

With CRS3 out the door, we can now start to work on new problems and the non-Latin characters is becoming a bigger issue now standing in the way of wider adoption of the CRS.

As can be seen in closed issues #143 and #151, some of the problems are gone with CRS3 in the default installation. So probably the situation is not as bad as it used to be.

But the situation is now fairly unclear and what we need is a couple of test requests, that trigger alerts, namely in the default installation. We can then work on these test requests. If you guys could provide these, then please open them in a separate new issue. For now, I am closing this.

**Comment by csanders-git** [comment]
@dune73  do we have a master ticket for this problem?

**Comment by dune73** [comment]
No, I do not think so. But the that would be worthwhile. Or even a label.
